doc/OVERVIEW.md

```diff
@@ -0,0 +1,469 @@
+
+There are two concepts for Cross-VM communications called *Shred Data* and *Shared Queue*.
+
+We also recommend you watch [this talk](https://www.youtube.com/watch?v=XdWmm_mtVXI&t=1168s) for introduction.
+
+## *Shared Data (Shared KVS)*
+
+What if you want to have global request counters across all the Wasm VMs running in multiple worker threads? Or what if you want to cache some data that should be used by all of your Wasm VMs? Then *Shared Data* or equivalently *Shared KVS* will come into play.
+
+*Shared Data* is basically a key-value store that is shared across all the VMs (i.e. cross-VM or cross-threads). A shared-data KVS is created per [`vm_id`](#envoy-configuration) specified in the `vm_config`. That means you can share a key-value store across all Wasm VMs not necessarily with the same binary (`vm_config.code`). The only requirement is having the same `vm_id`.
+
+<img src="./images/shared_data.png" width="600">
+
+In the diagram above, you see that VMs with "vm_id=foo" share the same shared data strage even though they have different binary (hello.wasm and bye.wasm).
+
+Here's the shared data related API of this Go SDK in [hostcall.go](../proxywasm/hostcall.go):
+
+```go
+// GetSharedData is used for retrieving the value for given "key".
+// Returned "cas" is be used for SetSharedData on that key for
+// thread-safe updates.
+func GetSharedData(key string) (value []byte, cas uint32, err error)
+
+// SetSharedData is used for seting key-value pairs in the shared data storage
+// which is defined per "vm_config.vm_id" in the hosts.
+//
+// ErrorStatusCasMismatch will be returned when a given CAS value is mismatched
+// with the current value. That indicates that other Wasm VMs has already succeeded
+// to set a value on the same key and the current CAS for the key is incremented.
+// Having retry logic in the face of this error is recommended.
+//
+// Setting cas = 0 will never return ErrorStatusCasMismatch and always succeed, but
+// it is not thread-safe, i.e. maybe another VM has already set the value
+// and the value you see is already different from the one stored by the time
+// when you call this function.
+func SetSharedData(key string, value []byte, cas uint32) error
+```
+
+
+The API is straightforward, but the important part is its thread-safeness or cross-VM-safeness with "cas" or [Compare-And-Swap](https://en.wikipedia.org/wiki/Compare-and-swap) value.
+
+Please refer to [an example](../examples/shared_data) for demonstration.
+
+## *Shared Queue*
+
+What if you want to aggregate metrics across all the Wasm VMs in parallel to request/response processing? Or what if you want to push some cross-VM aggregated information to a remote server? Now *Shared Queue* is here for you.
+
+*Shared Queue* is a FIFO(First-In-First-Out) queue created per a pair of `vm_id` and the name of the queue. And A *queue id* is assigned uniquely to the pair (vm_id, name) which is used for enqueue/dequeue operations.
```

Comment:
" per a pair of `vm_id` and the name of the queue" -- does that mean 1 shared queue per `vm_id`?

Reply:
or that each VM can create multiple shared queues using vm_id and the queue name?

Reply:
sorry this means latter; I will change the text to be clearer!